### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by kenwdelong** [graphite-project/docker-graphite-statsd#85]
I'm running [my fork](https://github.com/kenwdelong/docker-graphite-statsd) of this project on EC2, running on one of Amazon's T3.  I'm encountering what looks like memory leaks.
- Docker 18.06-1-ce
- Amazon Linux 2018-03

![image](https://user-images.githubusercontent.com/6744405/51923836-a286e600-23a0-11e9-93d4-1742e2a1fbd6.png)

Has anyone seen this?  Any ideas how to troubleshoot or fix?

**Comment by deniszh** [comment]
Default `carbon.conf` has `MAX_CACHE_SIZE = inf` setting, means that cache size can grow potentially indefinitely. You should limit it to some value, you can check `cache.size` metric to know how metrics do you have.

**Comment by deniszh** [comment]
If you using Grafana you can use this dashboard - https://grafana.com/dashboards/311

**Comment by kenwdelong** [comment]
Excellent, thanks for the quick response!  Here's the results from the dashboard, which confirms the finger pointing at carbon:

![image](https://user-images.githubusercontent.com/6744405/51925874-d106c000-23a4-11e9-83de-061686cc531f.png)

And indeed, my `MAX_CACHE_SIZE` still has the default setting.  I will adjust and report back here when the new evidence is in!

**Comment by kenwdelong** [comment]
Strange - looking at disk activity, carbon completely stopped writing data to disk (I/O went to zero) abruptly at 0:00 UTC.  The cache exploded (as above) at about 15:20 UTC. When I restarted the container (at about 17:20 UTC) to pick up the configuration as above, I lost all the data from 0:00 UTC.

**Comment by kenwdelong** [comment]
I set the `MAX_CACHE_SIZE` to 2 million, but now I'm losing all the data.  There have been no disk writes since just before 20:00 UTC.

![image](https://user-images.githubusercontent.com/6744405/51943947-a54cff80-23cf-11e9-820b-7b52dce64ceb.png)

I'm not sure what to look for next. Any help is appreciated.

**Comment by piotr1212** [comment]
Are you running the latest version? That error should have been fixed, twice already..

Also, why are you thinking it is a memory leak? Maybe your max_updates is too low, your disks are too slow or you are just sending too much metrics. The cache is actually a write buffer, if more comes in than comes out memory usage grows.

**Comment by deniszh** [comment]
@kenwdelong 
Accordng to fork it's 1.1.4, latest version is 1.1.5. Also check MAX_UPDATES_PER_SECOND, as @piotr1212 said - probably 1000 is too low, for SSD it can be 5-10K easily.

**Comment by kenwdelong** [comment]
Excellent, thanks for the suggestions. I will admit to being a newbie at tuning the Graphite stack. I think I found the issue.

When I added Grafana to the image, for some unfathomable reason, I put Grafana on port 80 and changed the Graphite web app to port 81.  This was leading to a bunch of errors in carbon.log like

`29/01/2019 19:00:02 :: [tagdb] Error tagging stats.timers.prod.api.prod-api-1-localdomain.performance.BenjiController.upper_99, stats.timers.prod.api.prod-api-2-localdomain.performance.RestController.count_99: Error response 404 from http://127.0.0.1:80/tags/tagMultiSeries
`

I switched the ports back to 80 for the Graphite web app and 81 for Grafana, and now it's gone 24h without the cache growing at all.
